### DISCUSSION TITLE
Separate agent only binary

### DISCUSSION THREAD

**Discussion by pratikbin** [prometheus/prometheus#11425] [General]
Hello

I was wondering is there any way to compile a separate binary containing only agent, so we don't have to carry a whole ~117MB Prometheus just to use agent.

**Comment by LeviHarrison** [comment]
The agent fundamentally uses the same code base as normal Prometheus (just slightly different functionality), so there isn't really a way to build an agent-only binary per se.

But, you can reduce binary size by using [service discovery plugins](https://github.com/prometheus/prometheus#service-discovery-plugins), which allow you to eliminate unused service discovery methods, which take up most of the binary.

**Reply by pratikbin** [reply]
> The agent fundamentally uses the same code base as normal Prometheus (just slightly different functionality), so there isn't really a way to build an agent-only binary per se.

I get that, but functionalities like alerting and all can be eliminated and only essentials things like scraping, storage, and remote write can be added. Digging into code.

> But, you can reduce binary size by using [service discovery plugins](https://github.com/prometheus/prometheus#service-discovery-plugins), which allow you to eliminate unused service discovery methods, which take up most of the binary.

👍

**Reply by LeviHarrison** [reply]
> I get that, but functionalities like alerting and all can be eliminated and only essentials things like scraping, storage, and remote write can be added. Digging into code.

I presume that there's very little code that could be completely removed, especially compared to the much larger footprint taken up by external libraries. To the point that it wouldn't be worth the hassle.
